ext/reflection/tests/ReflectionGenerator_isClosed_002.phpt

```diff
@@ -0,0 +1,42 @@
+--TEST--
+ReflectionGenerator::isClosed: ->valid() terminates an empty generator.
+--FILE--
+<?php
+
+function empty_generator() {
+    return;
+    yield;
+}
+
+$gens = [
+    (function() {
+        yield;
+    })(),
+    empty_generator(),
+];
+
+foreach ($gens as $gen) {
+    $ref = new ReflectionGenerator($gen);
+    var_dump($ref->isClosed());
+    var_dump($gen->valid());
+    var_dump($ref->isClosed());
+    try {
+        var_dump($ref->getExecutingLine());
+    } catch (\Exception $e) {
+        echo $e->getMessage(), PHP_EOL;
+    }
+
+    echo PHP_EOL;
+}
+
+?>
+--EXPECTF--
+bool(false)
+bool(true)
+bool(false)
+int(11)
```

Comment:
This line looks wrong, no? It should be 10 I think.

Reply:
It's confusing indeed. My understanding is that the generator stops once right at the beginning (i.e. where the opening brace is) and then right *after* each yield. Thus it reports the line of the closing brace.

```
<?php

function gen() {
    yield "xxx";
    var_dump(yield "yyy");

    echo "foo";
    echo "bar";
}

$gen = gen();
$ref = new ReflectionGenerator($gen);
var_dump($ref->getExecutingLine());

foreach ($gen as $yield) {
	var_dump($yield);
	var_dump($ref->getExecutingLine());
}

```

Reports:

```
int(4)
string(3) "xxx"
int(5)
string(3) "yyy"
int(5)
NULL
foobar
```

Note how the var_dump() is not actually executed right away, but only in the next iteration.

Reply:
But in the test, the generator is never actually `next()`ed, so how does it get past the `yield`?

Reply:
> But in the test, the generator is never actually `next()`ed, so how does it get past the `yield`?

That's the `valid()` gotcha that is tested here: `valid()` will start the generator machinery and thus it will immediately close an empty generator. The test serves to showcase how `Generator::valid()` is not an equivalent alternative to `ReflectionGenerator::isClosed()`, because it will cause a change in behavior, instead of just *observing* like Reflection is supposed to.

This effectively is issue #11591 that @nielsdos pointed out in: https://github.com/php/php-src/pull/14167#pullrequestreview-2064197448

Reply:
As discussed privately, `getExecutingLine()` will refer to the next executing op, which in this case is the implicit `return`,  because `valid()` implicitly starts the generator.